tests/tarp_test.py

```diff
@@ -0,0 +1,388 @@
+
+
+######################################################################
+## Check TARP with SBI
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("method", [SNPE])
+def test_batched_prepare_estimates(method, model="mdn"):
+    """Tests running inference and checking samples with tarp."""
+
+    num_dim = 2
+    prior = BoxUniform(-ones(num_dim), ones(num_dim))
+
+    num_simulations = 1000
+    max_num_epochs = 20
+    num_tarp_runs = 100
+
+    likelihood_shift = -1.0 * ones(num_dim)
+    likelihood_cov = 0.3 * eye(num_dim)
+
+    def simulator(theta):
+        return linear_gaussian(theta, likelihood_shift, likelihood_cov)
+
+    inferer = method(prior, show_progress_bars=False, density_estimator=model)
+
+    theta, x = simulate_for_sbi(simulator, prior, num_simulations)
+
+    _ = inferer.append_simulations(theta, x).train(
+        training_batch_size=100, max_num_epochs=max_num_epochs
+    )
+
+    posterior = inferer.build_posterior()
+    num_posterior_samples = 256
+    thetas = prior.sample((num_tarp_runs,))
+    xs = simulator(thetas)
+
+    samples = _infer_posterior_on_batch(xs, posterior, num_posterior_samples)
+
+    assert samples.shape != thetas.shape
+    assert samples.shape[1:] == thetas.shape
+    assert samples.shape[0] == num_posterior_samples
+
+    samples_ = _prepare_estimates(
+        xs, posterior, num_posterior_samples, infer_batch_size=32
+    )
+
+    assert samples_.shape != thetas.shape
+    assert samples_.shape[1:] == thetas.shape
+    assert samples_.shape[0] == num_posterior_samples
+    assert samples_.shape == samples.shape
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize("method", [SNPE])
+def test_consistent_run_tarp_results_with_posterior(method, model="mdn"):
+    """Tests running inference and checking samples with tarp."""
+
+    num_dim = 2
+    prior = BoxUniform(-ones(num_dim), ones(num_dim))
+
+    num_simulations = 1000
+    max_num_epochs = 20
+    num_tarp_sims = 100
+
+    likelihood_shift = -1.0 * ones(num_dim)
+    likelihood_cov = 0.3 * eye(num_dim)
+
+    def simulator(theta):
+        return linear_gaussian(theta, likelihood_shift, likelihood_cov)
+
+    inferer = method(prior, show_progress_bars=False, density_estimator=model)
+
+    theta, x = simulate_for_sbi(simulator, prior, num_simulations)
+
+    _ = inferer.append_simulations(theta, x).train(
+        training_batch_size=100, max_num_epochs=max_num_epochs
+    )
+
+    posterior = inferer.build_posterior()
+    num_posterior_samples = 256
+
+    thetas = prior.sample((num_tarp_sims,))
+    xs = simulator(thetas)
+
+    ecp, alpha = run_tarp(
+        thetas,
+        xs,
+        posterior=posterior,
+        num_posterior_samples=num_posterior_samples,
+        num_bins=30,
+        do_norm=True,
+        rng_seed=41,
+    )
+
+    atc, kspvals = check_tarp(ecp, alpha)
+    print(atc, kspvals)
+    assert atc < 0.5
+    assert atc > -0.5
+    assert kspvals > 0.05
```

Comment:
The test is failing for me locally. I had to increase the number of simulations to 6000 and remove the max num epoch so that training can converge. Can you confirm this? 
```suggestion
    num_dim = 2
    prior = BoxUniform(-ones(num_dim), ones(num_dim))

    num_simulations = 6000
    num_tarp_sims = 1000
    num_posterior_samples = 1000

    likelihood_shift = -1.0 * ones(num_dim)
    likelihood_cov = 0.3 * eye(num_dim)

    def simulator(theta):
        return linear_gaussian(theta, likelihood_shift, likelihood_cov)

    inferer = method(prior, show_progress_bars=True, density_estimator=model)

    theta = prior.sample((num_simulations,))
    x = simulator(theta)

    _ = inferer.append_simulations(theta, x).train(training_batch_size=1000)

    posterior = inferer.build_posterior()

    thetas = prior.sample((num_tarp_sims,))
    xs = simulator(thetas)

    ecp, alpha = run_tarp(
        thetas,
        xs,
        posterior=posterior,
        num_posterior_samples=num_posterior_samples,
        num_bins=30,
        do_norm=True,
        rng_seed=41,
    )

    atc, kspvals = check_tarp(ecp, alpha)
    print(atc, kspvals)
    assert -0.5 < atc < 0.5
    assert kspvals > 0.05
```